fondant/dataset.py

```diff
@@ -166,82 +167,123 @@ def upload(self, save_path):
+        """Abstract method that returns the dataset manifest"""
 
+    @abstractmethod
+    def _add_and_parse_args(self) -> argparse.Namespace:
+        """Abstract method that adds and parses the component arguments"""
+
+    def get_component_arguments(self) -> t.Mapping[str, Argument]:
+        """
+        Dictionary representation of the input and output arguments of a component
         Returns:
-            Manifest: the output manifest
+            Input and output arguments of the component.
         """
-        # step 1: add and parse arguments
-        args = self._add_and_parse_args()
-        # step 2: create Fondant dataset based on input manifest
-        metadata = json.loads(args.metadata)
-        if self.type == "load":
-            # TODO ideally get rid of args.metadata
-            # by including them in the storage args,
-            # getting run_id based on args.output_manifest_path
-            manifest = Manifest.create(
-                base_path=metadata["base_path"],
-                run_id=metadata["run_id"],
-                component_id=metadata["component_id"],
-            )
-        else:
-            manifest = Manifest.from_file(args.input_manifest_path)
-        dataset = FondantDataset(manifest)
-        # step 3: load or transform data
-        if self.type == "load":
-            df = self.load(args)
-            dataset.add_index(df)
-            dataset.add_subsets(df, self.spec)
-        else:
-            # create dataframe, based on component spec
-            df = dataset.load_data(self.spec)
-            # provide this dataframe to the user
-            df = self.transform(
-                df=df,
-                args=args,
-            )
-            # TODO update index, potentially add new subsets
+        component_arguments = {}
+        kubeflow_component_spec = self.spec.kubeflow_specification
+        component_arguments.update(kubeflow_component_spec.input_arguments)
+        component_arguments.update(kubeflow_component_spec.output_arguments)
+        return component_arguments
 
-        # step 4: create output manifest
-        output_manifest = dataset.upload(save_path=args.output_manifest_path)
+    @abstractmethod
+    def run(self):
+        """Abstract method for running component"""
 
-        return output_manifest
 
-    def _add_and_parse_args(self):
-        """
-        Add and parse component arguments based on the component specification.
-        """
+class FondantLoadComponent(FondantComponent):
+    """Abstract base class for a Fondant loader component"""
+
+    def _add_and_parse_args(self) -> argparse.Namespace:
         parser = argparse.ArgumentParser()
+        component_arguments = self.get_component_arguments()
 
-        kubeflow_component = self.spec.kubeflow_specification
+        for arg in component_arguments.values():
+            # Input manifest is not required for loading component
+            if arg.name == "input_manifest_path":
```

Comment:
I would expect this arg to not be defined for a Load component instead of having to handle it specifically here. Without this, we can lift this method completely to the base class.

Reply:
Fair point, can we just then keep it as Optional by default for both components. Otherwise we might need to rework the `KubeflowComponentSpec` to have different specification for both loader and transform. https://github.com/ml6team/fondant/blob/main/fondant/component_spec.py#:~:text=class-,KubeflowComponentSpec,-%3A
@NielsRogge wondering what's your take on this?

Reply:
I feel like the user would only need to be exposed to a single `FondantComponent` class actually. We could figure out that the first component in the pipeline is the loading one and that all others are transform components. We can internally split up the kubeflow spec for loading vs transformation, indeed.

Reply:
Good point. I think we can eventually tackle but maybe in a separate PR.

Reply:
The interface for both components is different, since we need to provide an input dataframe for the transform components and not for the load components. I don't think it makes sense to try to align these completely.

Reply:
Ok, we can keep the separation also for the user in that case.